python/sglang/srt/entrypoints/openai/logprobs_processor.py

```diff
@@ -0,0 +1,144 @@
+from __future__ import annotations
+
+from typing import Any, Dict, List, Optional, Sequence
+
+from sglang.srt.entrypoints.openai.protocol import (
+    ChatCompletionTokenLogprob,
+    ChoiceLogprobs,
+    LogProbs,
+    TopLogprob,
+)
+
+
+class LogProbsProcessor:
+    """Factory converting raw token/top-logprobs to OpenAI-style ChoiceLogprobs."""
+
+    @classmethod
+    def from_response(cls, ret_item: Dict[str, Any]) -> ChoiceLogprobs:
+        meta = ret_item["meta_info"]
+        return cls._build(
+            output_token_logprobs=meta["output_token_logprobs"],
+            output_top_logprobs=meta.get("output_top_logprobs"),
+            streaming=False,
+        )
+
+    @classmethod
+    def from_streaming(cls, content: Dict[str, Any], n_prev: int) -> ChoiceLogprobs:
+        meta = content["meta_info"]
+        return cls._build(
+            output_token_logprobs=meta["output_token_logprobs"][n_prev:],
+            output_top_logprobs=(meta.get("output_top_logprobs") or [])[n_prev:],
+            streaming=True,
+        )
+
+    @classmethod
+    def from_completions(
+        cls,
+        *,
+        ret_item: Dict[str, Any],
+        input_token_logprobs: Sequence[Any] | None = None,
+        input_top_logprobs: Sequence[Any] | None = None,
+        n_prev: int = 0,
+    ) -> ChoiceLogprobs:
+        meta = ret_item["meta_info"]
+        return cls._build(
+            input_token_logprobs=input_token_logprobs,
+            input_top_logprobs=input_top_logprobs,
+            output_token_logprobs=meta["output_token_logprobs"][n_prev:],
+            output_top_logprobs=(meta.get("output_top_logprobs") or [])[n_prev:],
+            streaming=n_prev > 0,
+        )
+
+    @classmethod
+    def _build(
+        cls,
+        *,
+        output_token_logprobs: Sequence[Any],
+        output_top_logprobs: Optional[Sequence[Any]] = None,
+        input_token_logprobs: Optional[Sequence[Any]] = None,
+        input_top_logprobs: Optional[Sequence[Any]] = None,
+        streaming: bool,
+    ) -> ChoiceLogprobs:
+        logprobs = cls._to_openai_style(
+            input_token_logprobs=input_token_logprobs,
+            input_top_logprobs=input_top_logprobs,
+            output_token_logprobs=output_token_logprobs,
+            output_top_logprobs=output_top_logprobs,
+        )
+        token_logprobs_list = cls._process_tokens(
```

Comment:
If I am understanding the call flow correctly, it seems this function is getting called for both chat_completions and completions.
But I only see serving_chat.py need this function.